portal/src/components/roles-and-groups/list/UserDetailsScreenRoleListContainer.tsx

```diff
@@ -0,0 +1,186 @@
+        return;
+      }
+      const value = e.currentTarget.value;
+      setSearchKeyword(value);
+      // Reset offset when search keyword was changed.
+      setOffset(0);
+    },
+    []
+  );
+  const onClearSearchKeyword = useCallback(() => {
+    setSearchKeyword("");
+  }, []);
+
+  const [isAddRoleDialogHidden, setIsAddRoleDialogHidden] = useState(true);
+  const showAddRoleDialog = useCallback(
+    () => setIsAddRoleDialogHidden(false),
+    []
+  );
+  const hideAddRoleDialog = useCallback(
+    () => setIsAddRoleDialogHidden(true),
+    []
+  );
+
+  const groupRoles = useMemo(() => {
+    const groupsRolesTable: Record<string, UserRolesListItem> = {};
+    user.groups?.edges?.forEach((edge) => {
+      const group = edge?.node;
+      if (group?.roles?.edges == null) {
+        return;
+      }
+      group.roles.edges.forEach((roleEdge) => {
+        const role = roleEdge?.node;
+        if (role == null) {
+          return;
+        }
+        if (role.key in groupsRolesTable) {
+          groupsRolesTable[role.key].groups.push(group);
+        } else {
+          const roleWithGroups = {
+            ...role,
+            groups: [group],
+          };
+          groupsRolesTable[role.key] = roleWithGroups;
+        }
+      });
+    });
+
+    return Object.entries(groupsRolesTable).map(([_, value]) => value);
+  }, [user.groups?.edges]);
+
+  const userRoles: UserRolesListItem[] = useMemo(() => {
+    return (
+      user.roles?.edges?.flatMap((e) => {
+        if (e?.node) {
+          return [{ ...e.node, groups: [] }];
+        }
+        return [];
+      }) ?? []
+    );
+  }, [user.roles?.edges]);
+
+  const combinedRoles: UserRolesListItem[] = useMemo(() => {
+    return [...groupRoles, ...userRoles];
+  }, [groupRoles, userRoles]);
+
+  const filteredCombinedRoles = useMemo(() => {
+    if (isSearch) {
+      return searchRoles(combinedRoles, searchKeyword);
+    }
+
+    return combinedRoles.slice(offset, offset + pageSize);
+  }, [isSearch, combinedRoles, offset, searchKeyword]);
+
+  if (error != null) {
+    return <ShowError error={error} onRetry={refetch} />;
+  }
+
+  if (loading) {
+    return <ShowLoading />;
+  }
+
+  const totalCount = rolesQueryData?.roles?.totalCount ?? 0;
+
+  if (totalCount === 0) {
+    return <RolesEmptyView />;
+  }
+
+  return (
+    <>
+      <section className={cn("flex flex-col h-full", className)}>
+        <header className="flex flex-row items-center justify-between mb-8">
+          <SearchBox
+            className="max-w-[300px] min-w-0 flex-1 mr-2"
+            placeholder={renderToString("search")}
+            value={searchKeyword}
+            onChange={onChangeSearchKeyword}
+            onClear={onClearSearchKeyword}
+          />
+          <PrimaryButton
+            text={<FormattedMessage id="UserDetailsScreen.roles.add" />}
+            onClick={showAddRoleDialog}
+          />
+        </header>
+        <UserRolesList
+          className="flex-1 min-h-0"
```

Comment:
Same as the groups list in the previous pr,
1. Give it a minimum height of 200px to avoid it disappearing in short screens.
1. It should be `flex-1-0-auto` so that it will not be shorter than 200px.